managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java

```diff
@@ -256,6 +276,132 @@ public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boole
         }
     }
 
+    @AllArgsConstructor
+    private static final class ReadEntriesCallbackWithContext {
+        final ReadEntriesCallback callback;
+        final Object ctx;
+        final long startEntry;
+        final long endEntry;
+    }
+
+    private CachedPendingRead findBestCandidate(PendingReadKey key, Map<PendingReadKey, CachedPendingRead> ledgerCache,
+                                                AtomicBoolean created) {
+        synchronized (ledgerCache) {
+            CachedPendingRead existing = ledgerCache.get(key);
+            if (existing != null) {
+                return existing;
+            }
+            // TODO: use a sorted map
```

Comment:
I am not 100% that we need a SortedMap to make the search more efficient.
sorting the map by the "startEntry" does not help much.

in my tests it is more likely that the "endEntry" is in common (because during tailing reads you hit the wall of the last written message)